packages/aws-cdk-lib/core/lib/removal-policies.ts

```diff
@@ -0,0 +1,154 @@
+import { IConstruct } from 'constructs';
+import { Aspects, IAspect } from './aspect';
+import { CfnResource } from './cfn-resource';
+import { RemovalPolicy } from './removal-policy';
+
+/**
+ * Properties for applying a removal policy
+ */
+export interface RemovalPolicyProps {
+  /**
+   * Apply the removal policy only to specific resource types.
+   * Can be a CloudFormation resource type string (e.g., 'AWS::S3::Bucket').
+   * @default - apply to all resources
+   */
+  readonly applyToResourceTypes?: string[];
+
+  /**
+   * Exclude specific resource types from the removal policy.
+   * Can be a CloudFormation resource type string (e.g., 'AWS::S3::Bucket').
+   * @default - no exclusions
+   */
+  readonly excludeResourceTypes?: string[];
+
+  /**
+   * If true, overwrite any user-specified removal policy that has been previously set.
+   * This means even if the user has already called `applyRemovalPolicy()` on the resource,
+   * this aspect will override it.
+   * @default false - do not overwrite user-specified policies
+   */
+  readonly overwrite?: boolean;
+}
+
+/**
+ * The RemovalPolicyAspect handles applying a removal policy to resources
+ */
+class RemovalPolicyAspect implements IAspect {
+  constructor(
+    private readonly policy: RemovalPolicy,
+    private readonly props: RemovalPolicyProps = {},
+  ) {}
+
+  /**
+   * Checks if the given resource type matches any of the patterns
+   */
+  private resourceTypeMatchesPatterns(resourceType: string, patterns?: string[]): boolean {
+    if (!patterns || patterns.length === 0) {
+      return false;
+    }
+
+    return patterns.includes(resourceType);
+  }
+
+  public visit(node: IConstruct): void {
+    if (!CfnResource.isCfnResource(node)) {
+      return;
+    }
+
+    const cfnResource = node as CfnResource;
+    const resourceType = cfnResource.cfnResourceType;
+
+    const userAlreadySetPolicy =
+      cfnResource.cfnOptions.deletionPolicy !== undefined ||
+      cfnResource.cfnOptions.updateReplacePolicy !== undefined;
+
+    if (!this.props.overwrite && userAlreadySetPolicy) {
+      return;
+    }
+
+    if (this.resourceTypeMatchesPatterns(resourceType, this.props.excludeResourceTypes)) {
+      return;
+    }
+
+    if (
+      this.props.applyToResourceTypes?.length &&
+      !this.resourceTypeMatchesPatterns(resourceType, this.props.applyToResourceTypes)
+    ) {
+      return;
+    }
+
+    // Apply the removal policy
+    try {
+      cfnResource.applyRemovalPolicy(this.policy);
+    } catch (error) {
+      // Check if the error is an instance of the built-in Error class
+      if (error instanceof Error) {
+        throw new Error(`Failed to apply removal policy to resource type ${resourceType}: ${error.message}`);
+      } else {
+        // If it's not an Error instance, convert it to a string for the message
+        throw new Error(`Failed to apply removal policy to resource type ${resourceType}: ${String(error)}`);
+      }
```

Comment:
Is this try/catch necessary? Are there any (bad) effects of not having it?
(IMO,  I don't think we need this one.)

Reply:
I thought about it, and it might not be necessary, so I removed it.